pkg/agent/flowexporter/exporter/exporter.go

```diff
@@ -296,21 +292,27 @@ func (exp *flowExporter) sendFlowRecords() error {
 				if _, err := exp.sendDataSet(); err != nil {
 					return err
 				}
-				exp.numDataSetsSent = exp.numDataSetsSent + 1
 			}
-			// If the connection is not present in the conntrack table anymore,
-			// then we delete the connection in connection store and record store.
-			// Please note that our record cache policy is that we keep the record in
-			// the record map as long as the corresponding connection is present in
-			// the conntrack table.
-			if !record.Conn.IsPresent {
-				klog.V(2).Infof("Deleting the inactive connection with key: %v from both connection map and record map", key)
-				if err := exp.connStore.DeleteConnectionByKey(key); err != nil {
-					return err
-				}
+			exp.numDataSetsSent = exp.numDataSetsSent + 1
+			if err := exp.connStore.SetConnAsExported(key); err != nil {
+				return err
+			}
+
+			if flowexporter.IsConnectionDying(&record.Conn) || !record.Conn.IsPresent {
+				// If the connection is in dying state, we will delete the flow records
+				// from records map.
+				klog.V(2).Infof("Deleting the inactive flow records with key: %v from record map", key)
 				if err := exp.flowRecords.DeleteFlowRecordWithoutLock(key); err != nil {
 					return err
 				}
+				// if the connection does not exist in conntrack table, we will delete it
+				// from connection map.
+				if !record.Conn.IsPresent {
+					klog.V(2).Infof("Deleting the inactive connection with key: %v from connection map", key)
+					if err := exp.connStore.DeleteConnectionByKey(key); err != nil {
```

Comment:
Will this delete all the connections that are in dying state? We may not reach here at all as flow records are already deleted for the connections that do not present in conntrack table. Why not do it in connections.go?

Reply:
I had a try on deleting connections only in `connections.go` and found that many of the connections will be deleted before flow records get exported, and these flow records cannot be exported because their connections are no longer updating. That's why I thought we need to delete connections at `exporter.go`. 
Yes you are right about dying connections for which we already deleted flow records. I added another variable in conn called `AreRecordsDeleted` to make sure the flow records are deleted before we delete the connections.